subscriptions-api.md

```diff
@@ -0,0 +1,613 @@
+ignored and the subscription manager assigns one instead.
+
+#### 3.2.4.2. Retrieving a Subscription
+
+The **Retrieve** operation MUST be supported by compliant Event Producers. It
+returns the specification of the identified subscription.
+
+Parameters:
+
+- id (string) - REQUIRED. Identifier of the subscription.
+
+Result:
+
+- subscription (subscription) - REQUIRED. Subscription object.
+
+Errors:
+
+- **ok** - the operation succeeded
+- **notfound** - a subscription with the given _id_ already exists
+
+#### 3.2.4.3. Querying for a list of Subscriptions
+
+The **Query** operation SHOULD be supported by compliant Event Producers. It
+allows to query the list of subscriptions on the subscription manager associated
+with or otherwise visible to the party making the request. If supported, it MUST
+be supported at the same endpoint as the **Create** subscription operation.
+
+Parameters:
+
+- none
+
+Result:
+
+- subscription (list of subscription) - REQUIRED. List of subscription objects
+
+Errors:
+
+- **ok** - the operation succeeded and returned results
+- **nocontent** - the operation succeeded and returned no results
+
+Protocol bindings and implementations of such bindings MAY add custom filter
+constraints and pagination arguments as parameters. A request without filtering
+constraints SHOULD return all available subscriptions associated with or
+otherwise visible to the party making the request.
+
+#### 3.2.4.4. Updating a Subscription
+
+The Update operation MAY be supported by compliant Event Producers. To request
+the update of a Subscription, the client submits a proposed subscription object
+whose _id_ MUST match an existing subscription. All other properties MAY differ
+from the original subscription. The subscription manager then updates the
+subscription and returns a subscription object that also contains all OPTIONAL
+properties for which default values have been applied.
+
+Parameters:
+
+- subscription (subscription) - REQUIRED. Proposed subscription object.
+
+Result:
+
+- subscription (subscription) - REQUIRED. Realized subscription object.
+
+Protocol bindings MAY map the Update and the Create operation into a composite
+"upsert" operation that creates a new subscription if one with the given _id_
+does not exist. In this case, the operation is \*_Create_ and follows that
+operation's rules.
+
+#### 3.2.8. Deleting a Subscription
+
+The **Delete** operation SHOULD be supported by compliant Event Producers. It
+returns the specification of the identified subscription.
+
+Parameters:
+
+- id (string) - REQUIRED. Identifier of the subscription.
+
+Result:
+
+- subscription (subscription) - REQUIRED. Subscription object.
+
+Errors:
+
+- **ok** - the operation succeeded
+- **notfound** - a subscription with the given _id_ already exists
+
+### 3.3. HTTP Binding for the Subscription API
+
+(TBD) This will be a straightforward mapping of the described API to a basic
```

Comment:
Just looking forward to the discussion about imperative vs declarative APIs... would it make sense to support a `202` + Location Header being returned for those implementations that can't commit to fully creating the subscription immediately?

Reply:
"Declarative API" is not a wire concept. You can certainly choose to have a runtime that is all driven by config files, but in the end, that runtime needs to go out and deal with actual resources. 

There 202 plus Location is weird. If you can already tell where you will create a resource and you're confident enough to return 2xx, you can just as well create it. 

I get the feeling that "Declarative API" is often code for "sorry, this is very slow".

Reply:
> I get the feeling that "Declarative API" is often code for "sorry, this is very slow".

LOL  I was just looking for a way to support both models. I agree that what appears on the wire should not change based on the server's impl choice except to say that in some cases an HTTP response might contain the created resource (or an indication that it completed the processing - e.g. a 201) while in other cases it might not due to async processing. And it's that async processing model that I consider the declarative impl sitting.

> There 202 plus Location is weird. If you can already tell where you will create a resource and you're confident enough to return 2xx, you can just as well create it.

if the processing of the request is done asynchronously, then the Location URL might not actually be the final URL, it might be a URL that sends a redirect to the real resource once it's created.  In HTTP terms, I view this more as the URL to the status of the create - which the http spec does talk about.

Reply:
That's an implementation choice. If you're confident that you're going to get the work done, it's ok to give a 201 + Location and then keep returning 307 + Retry-After for that URI until you're actually done. 

202 is a "got it, go away".

Reply:
> That's an implementation choice.

Right that's my point. If our spec allows both paths then we should be good. I was just thinking that if our spec says "just do it via http" but all of our examples, text, etc... only really talk about a "200" then I suspect a lot of client devs will only support that and will blow up on a 202+Location.